lib/datadog/core/configuration/settings.rb

```diff
@@ -431,36 +468,54 @@ def initialize(*_)
         # @default `DD_TAGS` environment variable (in the format `'tag1:value1,tag2:value2'`), otherwise `{}`
         # @return [Hash<String,String>]
         option :tags do |o|
-          o.default do
-            tags = {}
-
-            # Parse tags from environment
-            env_to_list(Core::Environment::Ext::ENV_TAGS, comma_separated_only: false).each do |tag|
-              key, value = tag.split(':', 2)
-              tags[key] = value if value && !value.empty?
-            end
+          o.env_var Core::Environment::Ext::ENV_TAGS
+          o.setter do |new_value, old_value|
```

Comment:
Why does `:tags` need its own custom setter now? Before it was happily using `env_to_list`..

Reply:
The `tags` is a complex setting. It accepts an ENV var that gets parsed as a list, and for customers configuring it directly though `Datadog.configured` block, it accepts a hash with the different tags. 

To keep the DSL simple and avoid extra complexity to support all cases. The settings that accept multiple value types do not have to define any `type` and push all the logic into the `setter` block. 

I'd like to explain this setting change in detail since it is the most complex one from the coed base.

Old code:
```ruby
option :tags do |o|
  o.default do
    tags = {}

    # Parse tags from environment
    env_to_list(Core::Environment::Ext::ENV_TAGS, comma_separated_only: false).each do |tag|
      key, value = tag.split(':', 2)
      tags[key] = value if value && !value.empty?
    end

    # Override tags if defined
    tags[Core::Environment::Ext::TAG_ENV] = env unless env.nil?
    tags[Core::Environment::Ext::TAG_VERSION] = version unless version.nil?

    tags
  end

  o.setter do |new_value, old_value|
    # Coerce keys to strings
    string_tags = new_value.collect { |k, v| [k.to_s, v] }.to_h

    # Cross-populate tag values with other settings
    if env.nil? && string_tags.key?(Core::Environment::Ext::TAG_ENV)
      self.env = string_tags[Core::Environment::Ext::TAG_ENV]
    end

    if version.nil? && string_tags.key?(Core::Environment::Ext::TAG_VERSION)
      self.version = string_tags[Core::Environment::Ext::TAG_VERSION]
    end

    if service_without_fallback.nil? && string_tags.key?(Core::Environment::Ext::TAG_SERVICE)
      self.service = string_tags[Core::Environment::Ext::TAG_SERVICE]
    end

    # Merge with previous tags
    (old_value || {}).merge(string_tags)
  end
end
```

We can see the `default`  block, parse the ENV to a list and transform its values into a hash, plus some overrides. This block is only executed if the value is not provided by `Datadog.configure` block.
The `setter`, in this case, always assumes that it receives a `hash` either because the ENV has been parsed in the `default` block or the customer has provided over `Datadog.configure` block.


New code:
```ruby
option :tags do |o|
  o.env_var Core::Environment::Ext::ENV_TAGS
  o.setter do |new_value, old_value|
    tag_list = case new_value
               when String
                 values = if new_value.include?(',')
                            new_value.split(',')
                          else
                            new_value.split(' ') # rubocop:disable Style/RedundantArgument
                          end

                 values.map! do |v|
                   v.gsub!(/\A[\s,]*|[\s,]*\Z/, '')

                   v.empty? ? nil : v
                 end

                 values.compact!
                 values.each_with_object({}) do |tag, tags|
                   key, value = tag.split(':', 2)
                   tags[key] = value if value && !value.empty?
                 end
               when Hash
                 new_value
               else
                 {}
               end

    env_value = env
    version_value = version
    service_name = service_without_fallback

    # Override tags if defined
    tag_list[Core::Environment::Ext::TAG_ENV] = env_value unless env_value.nil?
    tag_list[Core::Environment::Ext::TAG_VERSION] = version_value unless version_value.nil?

    # Coerce keys to strings
    string_tags = tag_list.collect { |k, v| [k.to_s, v] }.to_h

    # Cross-populate tag values with other settings
    if env.nil? && string_tags.key?(Core::Environment::Ext::TAG_ENV)
      self.env = string_tags[Core::Environment::Ext::TAG_ENV]
    end

    if version_value.nil? && string_tags.key?(Core::Environment::Ext::TAG_VERSION)
      self.version = string_tags[Core::Environment::Ext::TAG_VERSION]
    end

    if service_name.nil? && string_tags.key?(Core::Environment::Ext::TAG_SERVICE)
      self.service = string_tags[Core::Environment::Ext::TAG_SERVICE]
    end

    # Merge with previous tags
    (old_value || {}).merge(string_tags)
  end
end
```

We do not specify any `type` on the setting. The `setter` block is responsible for accounting for the case the value is a string coming from reading the ENV variable, and a `Hash` from `Datadog.configure` block. In the case of any of those, we default to an empty hash. 
Everything below that checking of type of value we are dealing with is identical as the old code 😄

Reply:
Makes sense!
I suggest leaving a comment saying we should deprecate allowing this to be with anything that does not reflect the environment variable String format (e.g. `c.tags = Hash`), as this will complicate long-term maintenance of this setting quite a lot.

Reply:
I decided to add `env_parser` block to account for complex env parsing logic. 

https://github.com/DataDog/dd-trace-rb/pull/2970/commits/d6ab986ab909ee6fabbcba4cb64d19107225bc65

Reply:
> As for having the parser as a block for env, no strong feelings

I think the decision of having the parser as an optional block argument for `env` is an interesting option, since it allows for better organisation, but could lead to repetition if we have a `env` and `deprecated_env`  option that has a custom parsing logic. 

```ruby
option :test do |o|
  o.env HELLO_WORLD do |value|
    custom logic 
  end
  o.deprecated_env HELLO_WORLD_OLD do |value|
    custom logic 
  end
end
```

Another point of view, is that it allow for individual parsing logic f the deprecated env var has a different format than new one for example. 

Another thing we can do (@ivoanjo might love it 😉) is remove `deprecated_env` and add support for defining multiple env + return the `env_var` to the `env` block. That would allow to move the complexity into the env block including the deprecation warning

```ruby
option :foo do |o|
  o.env ['NEW_ENV_VAR', 'OLD_ENV_VAR'] do |env_var, value|
    if env_var == 'OLD_ENV_VAR'
      # log deprecation
      # custom parsing logic if needed 
    else 
      # parse new env var
    end
  end
end
``` 

What do you guys think?